### PR TITLE
Update Presentation API proposal, open questions, and stakeholder feedback

### PR DESCRIPTION
**Pull request by takumif**

### CODE REVIEW

File: explainer.md [hunk 1]
```diff
@@ -42,6 +46,8 @@ secondary displays as well.
 - Allow the site to control whether to also capture the tab's audio while
   mirroring.
 - Allow the capture settings to be changed while mirroring is on-going.
+- Design the API in such a way that it can be extended to accommodate more
+  mirroring and non-mirroring configurations in the future.
```
Comment (markafoltz) [inline explainer.md]: Can you give an example of additional mirroring/non-mirroring configurations that we may want to support in the future?  Or is this referring to the ability to add more parameters to a presentation request URL?

File: explainer.md [hunk 2]
```diff
@@ -79,78 +85,75 @@ smart projector.
...
+let connection = null;
 
-// Define a PresentationRequest for mirroring.  Exact URL TBD.
-var request = new PresentationRequest("about:self");
-
-// Initial parameters for mirroring.
-var initialParams = { latencyHint: 'default', audioPlayback: 'remote' };
+const request = new PresentationRequest([{
+  // Mirroring URL, the exact URL TBD.
+  url: 'about:self',
```
Comment (markafoltz) [inline explainer.md]: The `about:` scheme is for specific browser internal pages, so I don't think an `about:` URL is quite right here.

The closest thing I know about in HTML is `_self`, which is the `target` to open links in the same tab.

File: explainer.md [hunk 3]
```diff
@@ -79,78 +85,75 @@ smart projector.
...
 
 ```
+partial interface PresentationRequest {
+  constructor(sequence<PresentationSource> sources);
+}
+
 partial interface PresentationConnection {
-  // Only non-null on a connection started by mirroring.
-  attribute CaptureParameters captureParams = null;
+  attribute PresentationSource source;
```
Comment (markafoltz) [inline explainer.md:151]: Will you deprecate the current `url` attribute?

File: explainer.md [hunk 4]
```diff
@@ -79,78 +85,75 @@ smart projector.
...
-  CaptureLatency latencyHint = 'default';
-  AudioPlaybackDestination audioPlayback = 'remote';
-};
+dictionary PresentationSource {
+  USVString url;
+  CaptureLatency? latencyHint = 'default';
+  AudioPlaybackDestination? audioPlayback = 'remote';
+}
 
 enum CaptureLatency { "default", "high", "low" };
```
Comment (markafoltz) [inline explainer.md]: Probably worth explaining what these values mean.

File: explainer.md [hunk 5]
```diff
@@ -161,50 +164,70 @@ enum AudioPlaybackDestination { "remote", "local" };
...
+1. Whether navigation in the tab, within the same origin or to a different
+   origin, should terminate the mirroring session.
 1. Whether this API should be restricted to top-level frames.
 1. Whether there's a use case to play audio out locally and remotely at the same
-time.
-1. Whether multiple URLs should be allowed in the PresentationRequest.
-1. The choice of token or URL to indicate self-mirroring as a presentation source.
-1. Whether captureParams should be available for other connection types (like 1-UA mode).
+   time.
+1. The choice of token or URL to indicate self-mirroring as a presentation
```
Comment (markafoltz) [inline explainer.md:186]: Let's go with `_self` for now.

**Review by markafoltz — APPROVED** [review]
A few minor comments, the new proposed URL is the only one that is really consequential.

File: explainer.md (hunk 1 above)
Comment (takumif) [reply explainer.md]: Pointed to the stakeholder feedback section for examples.

File: explainer.md (hunk 3 above)
Comment (takumif) [reply explainer.md:151]: Yes we can do that, I couldn't tell what the right deprecation syntax in WebIDL was so I used a made-up one.

File: explainer.md [hunk 6]
```diff
@@ -147,16 +149,29 @@ partial interface PresentationRequest {
 
 partial interface PresentationConnection {
   attribute PresentationSource source;
+  [deprecated] readonly attribute USVString url;
 };
 
 dictionary PresentationSource {
   USVString url;
```
Comment (markafoltz) [inline explainer.md]: This should be readonly

File: explainer.md [hunk 7]
```diff
@@ -147,16 +149,29 @@ partial interface PresentationRequest {
...
-  CaptureLatency? latencyHint = 'default';
-  AudioPlaybackDestination? audioPlayback = 'remote';
+  CaptureLatency? latencyHint = "default";
+  AudioPlaybackDestination? audioPlayback = "remote";
 }
 
-enum CaptureLatency { "default", "high", "low" };
+// Indicates the preferred tradeoff between low latency and smooth streaming.
+// The actual behavior is implementation specific. Only applies to mirroring
+// sessions.
```
Comment (markafoltz) [inline explainer.md]: s/sessions/sources/?

File: explainer.md [hunk 8]
```diff
@@ -147,16 +149,29 @@ partial interface PresentationRequest {
...
+  "default",
+  // Having a smooth stream is prioritized over having a low latency.
+  "high",
+  // Having a low latency is prioritized over having a smooth stream.
+  "low",
+};
 
+// Indicates where the audio playback should occur. Only applies to mirroring
+// sessions.
+// TODO: Should we use the terms "receiver" and "controller" instead?
```
Comment (markafoltz) [inline explainer.md]: Receiver/Controller would be more consistent with the current API.  I don't feel strongly either way though.

**Review by markafoltz — COMMENTED** [review]
LGTM

File: explainer.md (hunk 8 above)
Comment (takumif) [reply explainer.md]: Will use receiver/controller.

**Pull request merged**